### PR TITLE
[FEATURE array-helper] Implement array helper RFC #318

### LINKED ISSUE #17052: Implement + enable `array` Helper Feature
This is a tracking issue to enable the `array` helper as proposed in https://github.com/emberjs/rfcs/pull/318.

- [x] Implement
- [x] API Docs
- [x] Guides
- [x] Enable by default
- [ ] Author a polyfill (e.g. `ember-array-helper-polyfill`)?

### PR DESCRIPTION
**Pull request by josemarluedke**
This implements the array helper.
RFC: https://github.com/emberjs/rfcs/pull/318

Closes #17052.

### CODE REVIEW

File: packages/@ember/-internals/glimmer/lib/helpers/array.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+   ['Tom Date', 'Yehuda Katz', this.get('myOtherPerson')]
+   ```
+
+   Where the 3rd item in the array is bound to updates of the `myOtherPerson` property.
+
+   @method array
+   @for Ember.Templates.helpers
+   @param {Object} options
+   @return {Object} Array
+   @since 2.6.0
```
Comment (rwjblue) [inline packages/@ember/-internals/glimmer/lib/helpers/array.ts]: Seems wrong, possibly 3.7?

Also can you add `@category` + your feature flag to the docs, that lets us filter out API docs for unlanded features...

File: packages/@ember/-internals/glimmer/lib/helpers/array.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+   Use the `{{array}}` helper to create an array to pass as an option to your
+   components.
+
+   ```handlebars
+   {{my-component people=(array
+     'Tom Dade'
+     'Yehuda Katz'
+     myOtherPerson)
```
Comment (rwjblue) [inline packages/@ember/-internals/glimmer/lib/helpers/array.ts]: Lets use `this.myOtherPerson`

File: packages/@ember/-internals/glimmer/lib/helpers/array.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+   ```js
+   ['Tom Date', 'Yehuda Katz', this.get('myOtherPerson')]
+   ```
+
+   Where the 3rd item in the array is bound to updates of the `myOtherPerson` property.
+
+   @method array
+   @for Ember.Templates.helpers
+   @param {Object} options
+   @return {Object} Array
```
Comment (rwjblue) [inline packages/@ember/-internals/glimmer/lib/helpers/array.ts]: Should it be `@return {Array}`?

File: packages/@ember/-internals/glimmer/lib/helpers/array.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+   ```js
+   ['Tom Date', 'Yehuda Katz', this.get('myOtherPerson')]
+   ```
+
+   Where the 3rd item in the array is bound to updates of the `myOtherPerson` property.
+
+   @method array
+   @for Ember.Templates.helpers
+   @param {Object} options
```
Comment (rwjblue) [inline packages/@ember/-internals/glimmer/lib/helpers/array.ts]: I think we receive an array right?

File: packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    ['@test returns an array']() {
+      this.render(strip`{{#with (array \"Sergio\") as |people|}}
+                          {{#each people as |personName|}}
+                            {{personName}}
+                          {{/each}}
+                        {{/with}}`);
+
+      this.assertText('Sergio');
+
+      this.runTask(() => this.rerender());
```
Comment (rwjblue) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js]: I think we can use `this.assertStableRerender()`?

File: packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+      this.assertText('Sergio');
+    }
+
+    ['@test can have more than one value']() {
+      this.render(strip`{{#with (array \"Sergio\" \"Robert\") as |people|}}
```
Comment (rwjblue) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js]: given the template string, I don't think we need to escape the quotes

File: packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js [hunk 7]
```diff
@@ -0,0 +1,202 @@
+import { RenderingTest, moduleFor } from '../../utils/test-case';
+import { strip } from '../../utils/abstract-test-case';
+import { Component } from '../../utils/helpers';
+import { set } from '@ember/-internals/metal';
+
+moduleFor(
+  'Helpers test: {{array}}',
+  class extends RenderingTest {
+    ['@test returns an array']() {
+      this.render(strip`{{#with (array \"Sergio\") as |people|}}
```
Comment (rwjblue) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js]: Probably don't need to escape the quotes

File: packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+          personOne: 'Tom',
+          personTwo: 'Yehuda',
+        }
+      );
+
+      this.assertText('Tom,Yehuda,');
+
+      this.runTask(() => this.rerender());
+
+      this.assertText('Tom,Yehuda,');
```
Comment (rwjblue) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js]: Can probably replace the runTask and assertText with `this.assertStableRerender()`

File: packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+          personOne: 'Tom',
+          personTwo: 'Yehuda',
+        }
+      );
+
+      this.assertText('List:Tom,Yehuda,');
+
+      this.runTask(() => this.rerender());
+
+      this.assertText('List:Tom,Yehuda,');
```
Comment (rwjblue) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js]: Same as above, RE: stable re-render

**Review by rwjblue — COMMENTED** [review]
Looking good! I left some pretty nit-picky comments inline (sorry about that), but also would like to see a few more tests:

* Confirming that passing an `array` result into a component invocation, and _that_ component using `{{#each` properly rerenders/updates as items change
* Checking that the output of `array` follows the "immutable" style by having a component that we control which is passed an `array` and we inspect it for `===` equality.

File: packages/@ember/-internals/glimmer/lib/helpers/array.ts (hunk 2 above)
Comment (josemarluedke) [reply packages/@ember/-internals/glimmer/lib/helpers/array.ts]: I was wondering about that. Should the templates in the test also refer to `this` where there is a backing component? Also the `this.render` calls, are the second parameter considered arguments? I could change the tests to use `this.` and also `@`.

**Review by rwjblue — APPROVED** [review]

File: packages/@ember/-internals/glimmer/lib/helpers/array.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+   ```
+
+   Where the 3rd item in the array is bound to updates of the `myOtherPerson` property.
+
+   @method array
+   @for Ember.Templates.helpers
+   @param {Array} options
+   @return {Array} Array
+   @category array-helper
+   @since 2.7.0
```
Comment (btecu) [inline packages/@ember/-internals/glimmer/lib/helpers/array.ts]: This should be `3.6.0` at the earliest

File: packages/@ember/-internals/glimmer/lib/helpers/array.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+   @method array
+   @for Ember.Templates.helpers
+   @param {Array} options
+   @return {Array} Array
+   @category array-helper
+   @since 3.7.0
+   @public
+ */
+
+export default function(_vm: VM, args: Arguments) {
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/lib/helpers/array.ts]: Need return type

**Comment by chancancode** [comment]
Should we debug freeze the array? The empty case is frozen like hashes iirc

File: packages/@ember/-internals/glimmer/lib/helpers/array.ts (hunk 11 above)
Comment (josemarluedke) [reply packages/@ember/-internals/glimmer/lib/helpers/array.ts]: Would that be `any[]`?

Comment (josemarluedke) [reply packages/@ember/-internals/glimmer/lib/helpers/array.ts]: It seems seem to like that change alone.

screen shot 2018-10-10 at 11 00 56 am

The only way I found to make this work is by casting to any.

```ts
export default function(_vm: VM, args: Arguments): any[] {
  return args.positional.capture() as any;
}
```

**Comment by rwjblue** [comment]
> Should we debug freeze the array? The empty case is frozen like hashes iirc

Yes, probably.

File: packages/@ember/-internals/glimmer/lib/helpers/array.ts (hunk 11 above)
Comment (chancancode) [reply packages/@ember/-internals/glimmer/lib/helpers/array.ts]: I believe it should be `PathReference<Opaque>`, or `PathReference<Opaque[]>` if we want to be precise

**Comment by chancancode** [comment]
Unfortunately debug freezing is not going to be super easy. We can merge this as-is and do a follow up PR for that. I also just realized that the current version of the glimmer code seems to always return a new array/object even for the empty case.

We can do something like:

```
class DebugFreezeReference<T> implements PathReference<T> {
  constructor(private inner: PathReference<T>) {
    this.tag = inner.tag;
  }

  value(): T {
    return Object.freeze(this.inner.value());
  }

  get(path: string): PathReference<Opaque> {
    return this.inner.get(path); // or should this be `new DebugFreezeReference(this.inner.get(path))`?
  }
}
```

File: packages/@ember/-internals/glimmer/lib/helpers/array.ts (hunk 11 above)
Comment (chancancode) [reply packages/@ember/-internals/glimmer/lib/helpers/array.ts]: Can you also fix `hash.ts` to have a return type of `PathReference<Opaque>`? (or `PathReference<Dict<Opaque>>`)

Comment (josemarluedke) [reply packages/@ember/-internals/glimmer/lib/helpers/array.ts]: Pushed the change. I will open a new PR for the return type of hash helper to keep it separated.

File: packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+
+moduleFor(
+  'Helpers test: {{array}}',
+  class extends RenderingTest {
+    ['@test returns an array']() {
+      this.render(strip`{{#with (array "Sergio") as |people|}}
+                          {{#each people as |personName|}}
+                            {{personName}}
+                          {{/each}}
+                        {{/with}}`);
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js]: does the linter complain if you do something like...

```
this.render(strip`
  {{#with (array "Sergio") as |people|}}
    {{#each people as |personName|}}
      {{personName}}
    {{/each}}
  {{/with}}`
);
```

File: packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+        }
+      );
+
+      this.assertText('Tom');
+
+      this.assertStableRerender();
+
+      this.runTask(() => set(this.context, 'personOne', 'Yehuda'));
+      this.assertText('Yehuda');
+    }
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js]: Add another step to set it back to Tom (see #13127)

File: packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+      this.assertStableRerender();
+
+      this.runTask(() => set(this.context, 'personOne', 'Sergio'));
+
+      this.assertText('Sergio,Yehuda,');
+
+      this.runTask(() => set(this.context, 'personTwo', 'Tom'));
+
+      this.assertText('Sergio,Tom,');
+    }
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js]: add another step to set it back to Tom/Yehuda (see #13127)

File: packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+
+      this.assertStableRerender();
+
+      this.runTask(() => set(this.context, 'personOne', 'Chad'));
+
+      this.assertText('List:Chad,Yehuda,');
+
+      this.runTask(() => set(this.context, 'personTwo', 'Balint'));
+
+      this.assertText('List:Chad,Balint,');
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js]: add reset step

File: packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+
+      this.assertStableRerender();
+
+      this.runTask(() => set(fooBarInstance, 'model.personOne', 'Godfrey'));
+
+      this.assertText('Godfrey');
+
+      this.runTask(() => set(fooBarInstance, 'model', { personOne: 'Chad' }));
+
+      this.assertText('Chad');
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js]: add reset step

File: packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+      this.render(strip`{{foo-bar people=(array "Tom" personTwo)}}`, { personTwo: 'Chad' });
+
+      this.assertText('Tom,Chad,');
+
+      this.assertStableRerender();
+
+      this.runTask(() => set(this.context, 'personTwo', 'Godfrey'));
+
+      this.assertText('Tom,Godfrey,');
+    }
```
Comment (chancancode) [inline packages/@ember/-internals/glimmer/tests/integration/helpers/array-test.js]: add reset step

**Comment by chancancode** [comment]
Looking good. Probably should have a few more test testing "capturing" the array into JS values and make sure it is correct. Something like this:

```js
let captured;

this.registerHelper('capture', function([array]) {
  captured = array;
  return "captured";
});

this.render(`{{capture (array ...)}}`);

assert.deepEqual(captured, [...]);

this.runTask(this.set(...));

assert.deepEqual(captured, [...]);

...
```

Basically what you already did in the last test, but more coverage around the actual values, and that it updates correctly, etc. You can probably just merge that into your last test since it's already doing similar things.

**Comment by josemarluedke** [comment]
I have addressed the last few comments from @chancancode. This should be ready for another review.

**Review by chancancode — APPROVED** [review]

**Comment by chancancode** [comment]
@josemarluedke have you tried the indentation style in https://github.com/emberjs/ember.js/pull/17054#discussion_r224330568? If the linter prefer you write it like this, it's nbd

**Comment by josemarluedke** [comment]
@chancancode Yes I did, I changed most of the code to that indentation style. I did not change all because some already [had a line break due](https://github.com/emberjs/ember.js/pull/17054/files#diff-938074d5f22da2cef01e6e0e38d85f36R182) to component args. But if you do think I should change that as well, I'm happy to do so.

**Comment by rwjblue** [comment]
Thank you!

**Pull request merged**